### ISSUE TITLE
Debezium can inject snapshots in the middle of the stream, which causes data duplication on the Mz side

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#14211]
### What version of Materialize are you using?

510b6a9aa3da6253ca1f9773de9d27db9c07d57a

### How did you install Materialize?

Docker image

### What is the issue?

Apparently nothing prevents Debezium from re-snapshotting the entire table and inserting the snapshot again into the stream. Mz will then ingest that snapshot without retracting any data that existed previously in the dataflow. Wrong result will occur.

A scenario where a second snapshot is produced is described in https://issues.redhat.com/browse/DBZ-1420. Basically, restarting the source Postgres instance causes Debezium to restart its "connector". The connector then attempts to find a previously-saved offset, finds none and proceeds to perform a full snapshot and put it into the topic.

The sequence of events is as follows:

1. On the Pg side, do:
```
                CREATE TABLE debezium_table (f1 TEXT, f2 INTEGER, PRIMARY KEY (f1));
                ALTER TABLE debezium_table REPLICA IDENTITY FULL;
                INSERT INTO debezium_table SELECT 'A', 1;
```
2. Restart Pg
3. On the Pg side, do:
```
                UPDATE debezium_table SET f2 = f2 + 1;
``` 

The debezium log will show two snapshots being made. The topic will show two events, each with `snapshot` = `last`,  whatever that is supposed to mean.

```
{"before":null,"after":{"postgres.public.debezium_table.Value":{"f1":"A","f2":{"int":1}}},"source":{"version":"1.9.5.Final","connector":"postgresql","name":"postgres","ts_ms":1660304398619,"snapshot":{"string":"last"},"db":"postgres","sequence":{"string":"[null,\"23493504\"]"},"schema":"public","table":"debezium_table","txId":{"long":560},"lsn":{"long":23493504},"xmin":null},"op":"r","ts_ms":{"long":1660304398621},"transaction":null}

{"before":null,"after":{"postgres.public.debezium_table.Value":{"f1":"A","f2":{"int":3}}},"source":{"version":"1.9.5.Final","connector":"postgresql","name":"postgres","ts_ms":1660304429383,"snapshot":{"string":"last"},"db":"postgres","sequence":{"string":"[null,\"23494456\"]"},"schema":"public","table":"debezium_table","txId":{"long":563},"lsn":{"long":23494456},"xmin":null},"op":"r","ts_ms":{"long":1660304429383},"transaction":null}
```

Notice that the two events are completely unique -- none of their identifiers match so they can not be deduplicated. Theoretically we could take advantage of the fact that `f1` is a primary key, but currently we do not seem to do.

The expected result is:

```
postgres=# select * from debezium_table;
 f1 | f2 
----+----
 A  |  4
```

but instead we return:

```
materialize=> select * from debezium_source1;
 f1 | f2 
----+----
 A  |  2
 A  |  4
```

**Comment by philip-stoev** [comment]
To reproduce, clone `https://github.com/philip-stoev/materialize/`, branch `gh14211` and run:

```
cd test/platform-checks
./mzcompose down -v ; ./mzcompose run default --scenario=RestartSourcePostgres  --check=DebeziumPostgres 
```

**Comment by benesch** [comment]
@philip-stoev can you confirm this happens only with `ENVELOPE DEBEZIUM` and not `ENVELOPE DEBEZIUM UPSERT`? If so, this is a known issue and is tracked by this work item in #11772: 

<img width="812" alt="image" src="https://user-images.githubusercontent.com/882976/184351237-cfc37056-da8a-4595-a97d-6a0badbc38a9.png">

**Comment by philip-stoev** [comment]
@benesch if the workload is not updating the primary key,  `ENVELOPE DEBEZIUM UPSERT` seems to work. However, both Debezium envelope types are equally affected and return wrong results if the workload updates the primary key.

It is actually worse than not having prior state in order to skip duplicates, but in a different way. If one updates the primary key of a record from one value to another, each of the two snapshots will contain insertions for a different PK ... I think the only solution would be, upon detecting a second snapshot in the stream, to issue retractions for all records in the source so that the source is emptied out and can properly  be populated by the second snapshot.

**Comment by benesch** [comment]
> However, both Debezium envelope types are equally affected and return wrong results if the workload updates the primary key.

Oh, what a total disaster. I don't know that this is even solvable. I don't think Debezium indicates in any particular way when it retries a snapshot. I mean, you might know better than I at this point! Have you seen any way to detect in the Kafka topic when Debezium has retried the snapshot?

**Comment by aljoscha** [comment]
Also looping in @brennanvincent (née umanwizard)

**Comment by philip-stoev** [comment]
The only solution that I can see is the following, but I do not really know. If a snapshot is received:
 - if the previous event is a non-snapshot one, assume that this is a new snapshot
 - if previous event is a snapshot, compare the LSNs to determine if they a new snapshot has happened or not.

This is all very fragile:
- the Debezium documentation around snapshots appears to lag the code. That is, the documentation says that a snapshot is marked as `"snapshot:": "true"` , but actually it is `"snapshot": "latest"` . The documentation does not say what "latest" would mean
- this decision will be local to the a topic, so I guess the source may become inconsistent with other Debezium sources that may or may not be affected by the restart of the connector
- Debezium has something called incremental snapshots, I do not know how those are supposed to work.

**Comment by brennanvincent** [comment]
@philip-stoev what do the debezium records look like if the PK is updated?

**Comment by philip-stoev** [comment]
@brennanvincent if part of the normal stream, a delete of the pre-update record followed by a fresh insert of the updated record.

However, if snapshots are involved, then snapshot #1 will contain an insertion of the pre-update record and snapshot #2 will contain an insertion of the post-update record.

**Comment by benesch** [comment]
> * if the previous event is a non-snapshot one, assume that this is a new snapshot

I don't *think* this can happen. Or if it can, it'll be _extremely_ rare. Once Debezium emits a complete snapshot, it flips into replication mode and never emits another snapshot.

The only exception to this is the [incremental snapshot](https://debezium.io/blog/2021/10/07/incremental-snapshots/) feature. But those are pretty nonstandard, and require a separate table that signals to Debezium when you want to take a snapshot. AFAIK none of our users use this today.

> if previous event is a snapshot, compare the LSNs to determine if they a new snapshot has happened or not.

Last I checked, snapshots don't have LSNs. 🙈

**Comment by brennanvincent** [comment]
> I don't think this can happen.

@benesch what do you mean? Isn't that the entire issue we're discussing? Debezium can apparently re-snapshot under certain conditions.

**Comment by philip-stoev** [comment]
> Last I checked, snapshots don't have LSNs. see_no_evil

At least in the kafka messages I captured above, they seem to have some:

```
{"before":null,"after":{"postgres.public.debezium_table.Value":{"f1":"A","f2":{"int":1}}},"source":{"version":"1.9.5.Final","connector":"postgresql","name":"postgres","ts_ms":1660304398619,"snapshot":{"string":"last"},"db":"postgres","sequence":{"string":"[null,\"23493504\"]"},"schema":"public","table":"debezium_table","txId":{"long":560},"lsn":{"long":23493504},"xmin":null},"op":"r","ts_ms":{"long":1660304398621},"transaction":null}

{"before":null,"after":{"postgres.public.debezium_table.Value":{"f1":"A","f2":{"int":3}}},"source":{"version":"1.9.5.Final","connector":"postgresql","name":"postgres","ts_ms":1660304429383,"snapshot":{"string":"last"},"db":"postgres","sequence":{"string":"[null,\"23494456\"]"},"schema":"public","table":"debezium_table","txId":{"long":563},"lsn":{"long":23494456},"xmin":null},"op":"r","ts_ms":{"long":1660304429383},"transaction":null}
```

**Comment by brennanvincent** [comment]
Found something interesting from the Dbz docs:

> Primary key updates

An UPDATE operation that changes a row’s primary key field(s) is known as a primary key change. For a primary key change, in place of an UPDATE event record, the connector emits a DELETE event record for the old key and a CREATE event record for the new (updated) key. These events have the usual structure and content, and in addition, each one has a message header related to the primary key change:

    The DELETE event record has __debezium.newkey as a message header. The value of this header is the new primary key for the updated row.

    The CREATE event record has __debezium.oldkey as a message header. The value of this header is the previous (old) primary key that the updated row had.

**Comment by brennanvincent** [comment]
Tl;dr it seems you can read Kafka headers to detect PK changes.

**Comment by brennanvincent** [comment]
Also -- @philip-stoev when you said this:

> the Debezium documentation around snapshots appears to lag the code. That is, the documentation says that a snapshot is marked as "snapshot:": "true" , but actually it is "snapshot": "latest" . The documentation does not say what "latest" would mean

Did you mean to write `"last"`, instead of `"latest"` ? If so, I can explain what it means: `"last"` is supposed to represent the last record in the snapshot. `"true"` represents some record in the snapshot other than the last one. `"false"` means it is not part of the snapshot at all. See https://issues.redhat.com/browse/DBZ-1295 for the issue that introduced this.

I do not know what this is useful for or why they added it.

I also do not know why you are observing a snapshot with several `"last"` entries. That would seem to indicate a bug in Debezium if indeed that's what you're observing.

**Comment by philip-stoev** [comment]
> I also do not know why you are observing a snapshot with several "last" entries. That would seem to indicate a bug in Debezium if indeed that's what you're observing.

The examples above are two separate 1-record snapshots

**Comment by brennanvincent** [comment]
OK, got it. In that case, it makes sense for them to have snapshot=last.

**Comment by brennanvincent** [comment]
I need to check the dbz code to confirm this, but I think maybe you can detect new snapshots by checking whether the LSN has changed.

**Comment by brennanvincent** [comment]
I've [asked on their Zulip](https://debezium.zulipchat.com/#narrow/stream/302529-users/topic/Detecting.20new.20snapshots) to confirm. I think this will probably be quicker than trying to figure it out from the code.

**Comment by benesch** [comment]
Ooh, that's a fun edge case to consider if we end up reasoning about `snapshot=first` (forthcoming in Debezium 2.0). If the same record is both the first and last record in the snapshot... Debezium has to pick either "first" or "last", as it can't pick both. Oy.

@brennanvincent today in the office noted that he thinks the LSN can be used to detect when a snapshot has restarted. We _think_ the LSN field in the snapshot is the LSN at which the snapshot started. (That LSN has historically been useless for deduplicating on a record-by-record basis, but it's perfect for this, i.e., deduplicating one snapshot from another in its entirety.) @brennanvincent is double checking this with the Debezium folks on their Zulip.

**Comment by benesch** [comment]
Oh, sigh, GitHub didn't show me the updated comments. Anyway, thanks @brennanvincent!

**Comment by brennanvincent** [comment]
I've come around to @benesch 's view here which is that we should just not attempt to handle this until the Debezium folks give us a per-snapshot UUID to use. Trying to figure it out from metadata information like the LSN is too brittle and there are too many edge cases. They have an open issue for that, but it's not high-pri, so we would probably need to contribute upstream if we wanted it: https://issues.redhat.com/browse/DBZ-4338

I think for now we can simply flag this as an edge case and let customers know that they should check the Debezium logs to make sure it is not re-snapshotting.
